### PR TITLE
Adding a Copy Constructor to AudioStreamBuilder

### PR DESCRIPTION
**Pull request by atneya**
See issue #527 for more details.

### CODE REVIEW

File: samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp [hunk 1]
```diff
@@ -112,25 +83,7 @@ void PlayAudioEngine::createPlaybackStream() {
     }
 }
 
-/**
- * Sets the stream parameters which are specific to playback, including device id and the
- * callback class, which must be set for low latency playback.
- * @param builder The playback stream builder
- */
-void PlayAudioEngine::setupPlaybackStreamParameters(oboe::AudioStreamBuilder *builder) {
-    builder->setAudioApi(mAudioApi);
```
Comment (philburk) [inline samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp]: Did we lose the ability to set the AudioAPI?

File: samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp [hunk 2]
```diff
@@ -269,23 +215,16 @@ PlayAudioEngine::calculateCurrentOutputLatencyMillis(oboe::AudioStream *stream,
  * @see oboe::StreamCallback
  */
 void PlayAudioEngine::onErrorAfterClose(oboe::AudioStream *oboeStream, oboe::Result error) {
-    if (error == oboe::Result::ErrorDisconnected) restartStream();
+    if (error == oboe::Result::ErrorDisconnected) {
+        oboe::AudioStreamBuilder builder_ = oboe::AudioStreamBuilder(*oboeStream);
+        restartStream(&builder_);
```
Comment (philburk) [inline samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp]: Do we leak the old AudioStream pointer?
If you delete it then you may hit this issue!
https://github.com/google/oboe/issues/519

File: samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp (hunk 1 above)
Comment (atneya) [reply samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp]: We can still set the AudioAPI in  `PlayAudioEngine::setAudioApi`. This method is called when the user modifies the AudioAPI spinner in the interface. When the stream starts, it uses the default.

File: samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp (hunk 2 above)
Comment (atneya) [reply samples/hello-oboe/src/main/cpp/PlayAudioEngine.cpp]: Restarting will use the same pointer, `mPlayStream`, but create a new object. The old object is closed but not deleted, however, I think this issue should be addressed by using a unique pointer, which involves issue #529

File: samples/LiveEffect/src/main/cpp/FullDuplexStream.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * Called by Oboe when the stream is ready to process audio.
+     */
+    oboe::DataCallbackResult onAudioReady(
+            oboe::AudioStream *audioStream,
+            void *audioData,
+            int numFrames) override;
+
+    int32_t getMNumInputBurstsCushion() const;
```
Comment (philburk) [inline samples/LiveEffect/src/main/cpp/FullDuplexStream.h]: remove 'M' from getMNum and setMNum...

File: samples/LiveEffect/src/main/cpp/FullDuplexStream.h [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    /**
+     * Number of bursts to leave in the input buffer as a cushion.
+     * Typically 0 for latency measurements
+     * or 1 for glitch tests.
+     *
+     * @param mNumInputBurstsCushion
+     */
+    void setMNumInputBurstsCushion(int32_t mNumInputBurstsCushion);
```
Comment (philburk) [inline samples/LiveEffect/src/main/cpp/FullDuplexStream.h]: parameter should starts with "num"

File: samples/LiveEffect/src/main/cpp/FullDuplexStream.h [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+private:
+
+    // TODO add getters and setters
+    static constexpr int32_t kNumCallbacksToDrain   = 20;
+    static constexpr int32_t kNumCallbacksToDiscard = 30;
+
+    // let input fill back up, usually 0 or 1
+    int32_t mNumInputBurstsCushion = 1;
```
Comment (philburk) [inline samples/LiveEffect/src/main/cpp/FullDuplexStream.h]: align with other m variables

File: samples/LiveEffect/src/main/cpp/FullDuplexStream.h [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        return mOutputStream;
+    }
+
+    virtual oboe::Result start();
+
+    virtual oboe::Result stop();
+
+    /**
+     * Called when data is available on both streams.
+     * Caller should override this method.
```
Comment (philburk) [inline samples/LiveEffect/src/main/cpp/FullDuplexStream.h]: **App** should override

File: samples/LiveEffect/src/main/cpp/FullDuplexStream.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    virtual oboe::DataCallbackResult onBothStreamsReady(
+            const void *inputData,
+            int   numInputFrames,
+            void *outputData,
+            int   numOutputFrames
+            ) = 0;
+
+    /**
+     * Called by Oboe when the stream is ready to process audio.
+     */
```
Comment (philburk) [inline samples/LiveEffect/src/main/cpp/FullDuplexStream.h]: Add "This implements the stream synchronization. App should NOT override this method."

**Comment by atneya** [comment]
Do I need to roll back the rebase?

**Pull request merged**